src/plugins/rel/rel.js

```diff
@@ -140,11 +559,28 @@
             step.relative.z = 0;
         }
 
-        // Apply relative position to absolute position, if non-zero
-        // Note that at this point, the relative values contain a number value of pixels.
-        step.x = step.x + step.relative.x;
-        step.y = step.y + step.relative.y;
-        step.z = step.z + step.relative.z;
+        // Once rotate-x/rotate-y/rotate-z is set, all three must be set or use default 0
+        var useRelativeRotate = data.rotateX === undefined &&
+            data.rotateY === undefined &&
+            data.rotateZ === undefined;
+
+        if ( step.relative.position === "relative" ) {
+            var rel = rotateVector( worldUnitCoordinate, step.relative, prev.rotate );
+            step.x = step.x + rel.x;
+            step.y = step.y + rel.y;
+            step.z = step.z + rel.z;
+
+            if ( useRelativeRotate ) {
+                step.rotate = combineRotations( prev.rotate, step.relative.rotate );
```

Comment:
Why is this not just a sum like the above? What is all the math that follows good for?

Please add your reply as a code comment.

Reply:
The rotations in CSS is applied in order, and after each rotation, the rotation coordinate is updated accordingly. So we can't just sum up the angles. We need to know the new coordinate after each rotation, and calculate the position after the rotation according to it, and finally find out a one step rotation.

Reply:
Can you give some specific examples? I don't follow, and I just don't understand why this needs to be any more complicated than the example I've given in previous comment.

Reply:
As it's said in `examples/3D-rotations/index.html`, order is matter.

For example, in my firefox, the following two slides

```html
<div class="step"  data-rotate-x="90" data-rotate-z="90" data-rotate-order="xyz">XYZ</div>
<div class="step"  data-rotate-x="90" data-rotate-z="90" data-rotate-order="zyx">ZYX</div>
```

The XYZ one is horizontal and facing upside, the ZYX one is vertical and facing right side. 

In my opinion, the XYZ one is first rotated by X-axis, facing upward, and then rotated by the Z-axis of its rotated coordinate, which is the negative Y-axis before rotation.  So the result is just like two **independent** rotation, rotate-x=90 and rotate-y=-90.  The ZYX one is similar, the rotated X-axis is the original Y-axis, so it works like rotate-z=90 and then rotate-y=90.

So after each rotation, we have to know what the X/Y/Z-axis is pointing to, and do the following rotations by them. So we can not simply sum up angles of one axis.

For example, rotate-x=90 then rotate-z=90 then rotate-x=90 NOT equals to rotate-x=180 then rotate-z=90.

Hope I explained it clear.

Reply:
Furthermore, I suspect sum up the movement while `data-rel-position="relative"` is also problematic.  While `data-rel-position="absolute"`, `data-x` and `data-rel-x` are using the same axis, so it can be sum up. But while `data-rel-position="relative"`, `data-x` and `data-rel-x` may or maybe use the same axis, so it's not reasonable to be simply summed.

For example, `<div class="step" data-x="100" data-rel-y="100">`, what's user's purpose? 

If previous slide is at (0,0,0), and no rotation, the current slide will be x=100, y=100. If rotate-x=90 in previous slide, the effect will be x=0,y=0, just overlap with the previous one. If the user doesn't know the status of previous slide, how can he know the final status, how can he achieve what he need? If he knows, he should just use the absolute position.

So I think maybe we should disallow mixing up absolute and relative at all, while `data-rel-position="relative"`.

Reply:
Yes, order matters and that's why we have the ability to define rotate-order. This is how CSS works. It sounds like you're trying to work against this, and a) I don't quite understand why you're even doing this, and b) it leads to quite complicated code where you're essentially reverse engineering what CSS does in order to counter the rotation from normal CSS.

> For example, rotate-x=90 then rotate-z=90 then rotate-x=90 NOT equals to rotate-x=180 then rotate-z=90.

Correct, they are not the same. What I'm trying to say is that rotate-x=180 is the correct choice.

> If the user doesn't know the status of previous slide, how can he know the final status, how can he achieve what he need? If he knows, he should just use the absolute position.

Why would the user not know the position of the previous slide? It's the user who created the presentation.

The reason to use rel coordinates is not ignorance, but to save typing.

> So I think maybe we should disallow mixing up absolute and relative at all

Again, it really sounds like you are trying to do something different than what is the point of the rel plugin. The rel plugin for sure is meant to mix well on top of the absolute coordinates.

Reply:
> ﻿﻿Why would the user not know the position of the previous slide? It's the user who created the presentation.

For easy preparing presentation, users may split the presentation into several groups, design each group independently, then finally assemble them into the final presentation.  While they design one part, they should only care about the relative position, not the absolute position.  The absolute position will be determined at the latest stage, by changed the position of first slide in a group.

For example, in `examples/3D-positions/index.html`, the presentation is separated into three groups, one box and two rings.  Two rings have the same shape, have the same code except for the first slide in each group.  When I design them, I really don't know the final position of each slide.  And all I need in final is to alter the position of the first slide, to get what I want.

Reply:
> It sounds like you're trying to work against this, and a) I don't quite understand why you're even doing this, and b) it leads to quite complicated code where you're essentially reverse engineering what CSS does in order to counter the rotation from normal CSS.

All I want is let the computer do more, the user do less.

When I design a presentation using the originally version of `impress.js`, I met some problems.  I want to separate the presentation in to several parts, in each part, main slides are laid out in a line, slide in from right, and sub slides are laid out vertical below the main slide, slide up one by one.  It's easy, right?  But while I want to change the direction of slides, trouble comes.  E.g. when I want the three parts of slides form a triangle, slides in two of three parts should have an angle of ±60 degree. I found I have to carefully calculate the position of each slide, it is boring.   So I thought, why not just tell the computer, I want the next slide at the east side of current slide, and have a distance of 1w?  It's easy for computer to do this than we.

So I try finding a way to do this.  After some search, I found that quaternion is the most recommended way, so I used it.  It's a bit complicate, but I don't know whether there're simplier way.